### ISSUE TITLE
eslint error no-undef for looping variable

### ISSUE THREAD

**Issue by jconsigliere** [ezhlobo/eslint-plugin-react-pug#14]
Original issue can be found [here](https://github.com/pugjs/babel-plugin-transform-react-pug/issues/38).

**Comment by artificis** [comment]
@ezhlobo I tried to set up a new fresh repo to reproduce this behavior, but I couldn't make one 😃 
However, I'm certainly having this issue in one of my production codes (but can't share the whole repo, sorry). Let me spend a few more hours and I will post a link here.

**Comment by ezhlobo** [comment]
@artificis sure, thank you for reporting that ;). 

Can you set version of this plugin to `0.2.0` instead of `0.1.0` in your private project? I added supporting of `no-undef`, so it might affect on your mysterious issue.

**Comment by artificis** [comment]
Looks like there are some improvements, but I think I found one of the culprits - `no-undef` raises when I use attribute interpolation. E.g.,

```pug
each company in this.props.companies
  li(onClick=${() => this.handleCompanySelect(company)})
```

**Comment by artificis** [comment]
I know I can improve above code by using `bind` like this:
```pug
each company in this.props.companies
  li(onClick=this.handleCompanySelect.bind(this, company))
```
And `no-undef` error doesn't raise in this case. However, there are cases when we really need interpolations, `key` prop for instance,

```pug
each company, i in this.props.companies
  li(key=${`company-${i}`})
```
And this raises `no-undef` error for both `company` and `i`.

**Comment by ezhlobo** [comment]
@artificis hey, here is a lot of information. 

Your `no-undef` is coming from core of eslint, it's not related to pug. All your content within `${...}` inside template literals (inside backticks `) is a javascript and don't understand pug because it does not have access to it.

```js
pug`
  each company in this.props.companies
    li(onClick=${() => this.handleCompanySelect(company)})
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                 ^ this is javascript code which does not have access to pug template
`

pug`
  each company, i in this.props.companies
    li(key=${`company-${i}`})
             ^^^^^^^^^^^^^^
             ^ this is javascript code which does not have access to pug template
`
```

To make both of options properly you have to use inline scripts in pug template:

```js
pug`
  each company in this.props.companies
    - const handler = () => this.handleCompanySelect(company)
    li(onClick=handler)               
`

pug`
  each company, i in this.props.companies
    - const key = 'company-' + i
    li(key=key)
`
```

Let me know what do you think about it.

**Comment by artificis** [comment]
That makes sense in some ways, but it could look ugly with those lines (e.g., `- const key = 'company-' + i` to use variable only once?).
Also babel has no issue with transpiling the code when I disable eslint, so technically above "javascript code which does not have access to pug template" ran without any issue (all pug variables resolved to correct values)

**Comment by artificis** [comment]
And `li(key="company-#{i}")` didn't work as expected - it just put plain text "company-#{i}", not something like "company-1". I know this is not related to this repository here. It's more related to babel pug plugin. Just wanted to know if you have any ideas about why the @ForbesLindesay disabled this feature.

**Comment by ezhlobo** [comment]
@artificis 

>it could look ugly with those lines

I believe that we can think up better code if you provide me your context why do you care about value of key. It's supposed to be just unique so in your case I would pass just `i`. But if you construct `key` somehow, it should be done out of `render()` method in my opinion, because it's logic, not just representation.

>babel has no issue with transpiling the code when I disable eslint, so technically above "javascript code which does not have access to pug template" ran without any issue

It's something what babel plugins address, so it works eventually. But I agree that it should be addressed in eslint as well. I'm not sure that it's possible without many tradeoffs, but I'll do investigation.

>And li(key="company-#{i}") didn't work as expected

I think that it's a limitation from pug, not from that plugin: https://pugjs.org/language/attributes.html#attribute-interpolation

**Comment by artificis** [comment]
Ah okay. Thanks for clarifying. Do you want me to close out the issue or leave it open? (because I don't think I have any further thing to discuss for now)

**Comment by ezhlobo** [comment]
@artificis I'm going to rename it (if you don't mind) and try to figure out how to handle interpolation with eslint in pug.

**Comment by ForbesLindesay** [comment]
> And li(key="company-#{i}") didn't work as expected

I think you're used to a very old version of pug. We haven't supported that syntax for a couple of years now. Attribute values are just JavaScript. This means that both of the following **should** work, and if they don't, it's a bug.

```js
pug`
  each company in this.props.companies
    li(onClick=() => this.handleCompanySelect(company))               
`

pug`
  each company, i in this.props.companies
    li(key='company-' + i)
`
```

**Comment by artificis** [comment]
Yes, you are right. I learned that just now.

@ezhlobo Okay

**Comment by PhmNgocNghia** [comment]
can confirm this error have in interpolation mode

```
render () {
    return pug`
      Wrapper.d-flex.justify-content-center.align-items-center
        Card.p-5
          ModifiedDisplay.text-white.mb-4.text-center TodoList
          each authProvider, index in this.state.authProviders
            SocialButton(
              key=index
              className=${authProvider}
            ).btn.btn-block.btn-social.btn-lg.mb-4
              span(classNme='company-' + authProvider)
              | Continue with #{this.upperCaseFirst(authProvider)}

    `
  }
```

**Comment by ezhlobo** [comment]
@PhmNgocNghia this is because interpolation goes on javascript level. When you define variables in pug, they are on pugjs level. We can't merge these levels easily (you can send a PR with this and let me know if you need any help).

The best way to avoid that behavior is to avoid using interpolation. Like your example can be rewritten this way:

```diff
  Wrapper.d-flex.justify-content-center.align-items-center
    Card.p-5
      ModifiedDisplay.text-white.mb-4.text-center TodoList
      each authProvider, index in this.state.authProviders
        SocialButton(
          key=index
-         className=${authProvider}
+         className=authProvider
        ).btn.btn-block.btn-social.btn-lg.mb-4
          span(classNme='company-' + authProvider)
          | Continue with #{this.upperCaseFirst(authProvider)}
```

**Comment by PhmNgocNghia** [comment]
This maybe off-topic. To install eslint-plugin-react user must either eject or rewrite eslint config via create-react-wired. You should state it clear in the usage section.

**Comment by jconsigliere** [comment]
@PhmNgocNghia I didn't know about create-react-wired, but I've been using `react-app-rewired`. [Here](https://github.com/artificis/react-app-suspenders), you can see the example configuration of create-react-app using react-app-rewired.

**Comment by ezhlobo** [comment]
Thank you all for this conversation. I see there are two topics remained, here is an update:

## Updates:

* Interpolation should be restricted by eslint. Will create a rule for this. Related issue #30.
* We need to cover more how to use this plugin with create-react-app. This is something what related to [pugjs/babel-plugin-transform-react-pug](https://github.com/pugjs/babel-plugin-transform-react-pug) and will be handled there.
